V4L2 codec updates for interlaced decode

6by9 opened the pull request:
This puts all the correct signalling in place for interlaced content.

There is still a tweak needed so that the decoder stops on any change of colourspace or interlacing as required by the V4L2 spec, but FFmpeg doesn't appear to follow the spec in ALWAYS stopping and restarting the output (V4L2 CAPTURE) queue on a source change event, so it would most likely stall.

GStreamer 1.14.4 likewise seems to have an issue if you tell it that the output stream should be interlaced in that it tries to propagate it further up the pipeline, and h264parse seems to object. Telling it that it's progressive works fine. I'll try to sort a 1.19 build at some point, or Bullseye using 1.18, and see how that fairs.

@popcornmix It'd be interesting to know what Kodi makes of this change.

Chromium 88 seems to be happy with it (using BBC-News2017 clip from TestMedia), although it's a touch glitchy (1080i50). Nothing is actually deinterlacing the content of course.

6by9 committed:
staging/bcm2835-codec: Add support for decoding interlaced streams

The video decoder can support decoding interlaced streams, so add
the required plumbing to signal this correctly.

The encoder and ISP do NOT support interlaced data, so trying to
configure an interlaced format on those nodes will be rejected.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/vchiq-mmal: Add buffer flags for interlaced video

Add the buffer flags that the firmware uses to identify fields
on interlaced video

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/bcm2835-codec: Change the default codec res to 32x32

In order to effectively guarantee that a V4L2_EVENT_SOURCE_CHANGE
event occurs, adopt a default resolution of 32x32 so that it
is incredibly unlikely to be decoding a stream of that resolution
and therefore failing to note a "change" requiring the event.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/bcm2835-codec: Correct ENUM_FRAMESIZES stepsize to 2

Being YUV420 formats, the step size is always 2 to avoid part
chroma subsampling.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/vchiq-mmal: Add parameters for interlaced video support

Adds enum mmal_interlace_type and struct
mmal_parameter_video_interlace_type to allow for querying the
interlacing mode on decoders.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/bcm2835-codec: Return buffers to QUEUED not ERROR state

Should start_streaming fail, or buffers be queued during
stop_streaming, they should be returned to the core as QUEUED
and not (as currently) as ERROR.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
staging/bcm2835_codec: Log MMAL flags in hex

The flags is a bitmask, so it's far easier to interpret as hex
data instead of decimal.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

jc-kynesim commented:
I think last time I tried doing a "stream on" on every event we ended up up in an event loop with cropping, but that may be fixed now.  I'm happy to try again.

popcornmix commented:
@jc-kynesim are you happy with this?
I think it's okay from my testing. The issues I've seen have been in ffmpeg.

6by9 commented:
Rebased, and updated to assign to vb2->field instead of ```|=``.

6by9 committed:
staging: bcm2835-codec: Allow custom specified strides/bytesperline.

If the client provides a bytesperline value in try_fmt/s_fmt then
validate it and correct if necessary.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Added one extra patch that is currently untested that should allow a client specified bytesperline value for the stride.

popcornmix commented:
I think this can be merged. It allows interlaced flags to be reported to ffmpeg.
There will be a follow up PR with support for actual deinterlace.
